include/cudaq/Optimizer/CodeGen/Pipelines.h

```diff
@@ -20,29 +20,22 @@
 
 namespace cudaq::opt {
 
-// Pipeline builder to convert Quake to QIR.
-template <bool BaseProfile = false>
-void addPipelineToQIR(mlir::PassManager &pm) {
-  cudaq::opt::addAggressiveEarlyInlining(pm);
```

Comment:
I'm not sure removing a bulk of these passes is the right idea.

QIR profiles require/dictate specific characteristics about the code's control-flow, etc. Those characteristics are inherent to QIR but not necessarily some other backend. Removing these passes here, which are required for QIR, means that the QIR dependencies are just implied and expected to be handled elsewhere either by intention or accident.

The compiler should never break if a pass is run and has nothing to do. On the other hand, the output of the compiler can be non-conforming if required passes are accidentally never run.

It kind of defeats the point of a "pass pipeline" to strip a "big step" transformation back down to atomic passes someone has to (re)discover and orchestrate implicitly. A pass pipeline captures all the required passes that must be run and regardless of whatever passes may or may not have been run by other tools, pipelines, etc.

Reply:
While it may look like I was removing this code, it was actually effectively non-operational code to begin with, because the true backend code was never calling this function, so I would argue that I wasn't really removing _operational_ code here.

This particular pass pipeline is now called after the backend-specific mapping (like `ionq-gate-set-mapping`), as defined by the `PLATFORM_LOWERING_CONFIG` setting in the corresponding `.config` files, so bulk of the passes have already been done, and the only ones that remain are the ones listed here. On a related note, I suspect each particular backend needs to maintain the ability to have their own passes (as they currently do), and the prior implementation of this function didn't really support that (which is probably why it wasn't used in the operational code).

Reply:
Thanks. That helped clear things up a bit.

I'd like to see the following divisions made. If the output is to be QIR, a standard set of structural transformations will be required. Inlining call between kernels, unrolling of loops, etc. There should be a pass pipeline that is common to the runtime, cudaq-translate, etc. that encodes that part of the translation. The gate set mappings need to happen and those currently key off strings and information stuck in `.config` text files. Lastly, there is the pipeline to QIR and, possibly, one of the QIR profiles, which is what this captures. (Not all targets go to QIR, but I believe all targets require the structural transformations.)

With that step-wise evolution in mind, we should be able to define a template driver function that ties these steps together, though that can be done in a separate PR. (Probably requires reducing some of the exposed footprint in the `.config` files, which might not be an entirely bad idea.)